flashlight/fl/examples/RnnClassification.cpp

```diff
@@ -0,0 +1,260 @@
+    }
+    totalnumexamples += v.size();
+    std::cout << "Found " << v.size() << " examples for category " << lang << ". Total: " << totalnumexamples <<std::endl; 	
+    datasets[lang]=v;
+  }
+
+  // Turn a string into an AF array <1 x line_length> of char indices
+  static af::array lineToTensor(const std::string& line) {
+    std::vector<float> d;
+    for(char c : line) {
+      d.push_back((float)(c)); // direct cast of char to float
+    }
+    return af::array(1, d.size(), d.data());
+  }
+
+  ClassDataset(const std::string dataset_path) {
+    // Todo: search for all dataset files using std filesystem if cpp17 enabled
+    std::vector<std::string> lang = { "Arabic", "Greek", "Chinese", "Czech", "Dutch", "Japanese", "Korean", "Russian", "English", "Scottish", "Vietnamese", "German", "Spanish", "French", "Polish", "Italian", "Irish" };
+    for (auto& l : lang)
+      read(dataset_path, l);
+    for(auto& it : ID2label)
+      std::cout << it.first << ":" << it.second << ", ";
+    std::cout << std::endl;
+  }
+
+  // each epoch to go over some percent of the training dataset
+  int64_t size() const override { return .3f * totalnumexamples; }
+
+  // get a random example: name, category
+  std::pair<std::string, std::string> getRandomExample() const {
+    std::pair<std::string, std::string> p;
+    auto it=datasets.begin();
+    unsigned cat = randi(0, datasets.size()-1); // random category index
+    std::advance(it, cat);
+    p.second = it->first; // category name
+    const auto& v=it->second;
+    unsigned nn = v.size();
+    unsigned ri = randi(0,nn-1);
+    p.first = v[ri]; // name
+    FL_VLOG(fl::INFO) << "Random example:" << p.first << " class:" << cat << " : " << it->first;
+    return p;
+  }
+
+  // get a (random) example and return a vector of 2 tensors : the input and the expected category index
+  std::vector<af::array> get(const int64_t ) const override {
+    auto p = getRandomExample();
+    const std::string& n = p.first;
+    af::array input = lineToTensor(n);
+    std::vector<float> cv; 
+    cv.push_back(label2ID[p.second]);
+    af::array expected(1, cv.data());
+    return { input, expected };
+  }
+};
+std::map<std::string, unsigned> ClassDataset::label2ID;
+std::map<unsigned, std::string> ClassDataset::ID2label;
+
+class RnnClassifier : public Container {
+ public:
+  explicit RnnClassifier(unsigned num_classes, unsigned vocab_size, int hidden_size = 256, unsigned num_layers=2)
+      : embed(Embedding(hidden_size, vocab_size)),
+        rnn(RNN(hidden_size, hidden_size, num_layers, RnnMode::GRU, 0 /* Dropout */)),
+        linear(Linear(hidden_size, num_classes)),
+        logsoftmax(0) {
+    std::cout << "Creating a RNN Classifier with vocab size: " << vocab_size << " and num classes: " << num_classes << std::endl;
+    add(embed);
+    add(rnn);
+    add(linear);
+    add(LogSoftmax());
+  }
+
+  std::vector<Variable> forward(const std::vector<Variable>& inputs) override { return rnn(inputs); }
```

Comment:
why you don't call here std::tuple<Variable, Variable, Variable>
  forward(const Variable& input, const Variable& h, const Variable& c)? or vice versa, seems to return only rnn(inputs) is not correct forward?

Reply:
Sure, that's just what the RnnLm example is doing, not sure why. Anyway, not used in this example, so now disabled.